python/tank/platform/engine.py

```diff
@@ -1409,7 +1584,23 @@ def __load_apps(self):
                 for msg in messages:
                     self.log_warning("")
                     self.log_warning(msg)
-                
+
+            # Update the persistent application pool for use in context changes.
+            for app in self.__applications.values():
+                # We will only track apps that we know can handle a context
+                # change. Any that do not will not be treated as a persistent
+                # app.
+                if app.context_change_allowed:
+                    app_path = app.descriptor.get_path()
+
+                    if app_path not in self.__application_pool:
+                        self.__application_pool[app_path] = dict()
+
+                    self.__application_pool[app.descriptor.get_path()][app_instance_name] = app
+
+            # Update the persistent commands pool for use in context changes.
```

Comment:
I think it's worth explaining the caching strategy in a comment somewhere; that we have to cater for the use case where there are more than one instance of the same app in an environment and in this case we need to cache the same app more than once etc.